### ISSUE TITLE
Proposal: reuse test in CI for the release check

### ISSUE THREAD

**Issue by Yiyiyimu** [apache/apisix#3514]
### Background:

Currently, before we publish a release, there would be a call for the manual check for the release tarball in the email list, while we better should not rely on the manual check to find problems. CI could better serve this purpose.

### Github Actions Steps

1. The action would be triggered when the PR is (1) not in master branch (2) with tag `release-test`
2. Upload the package tarball to the PR, since the test step would be before the release so we need to do that ourselves. The file would be less than 200k for now, so it is tolerable for GitHub.
3. Get the file with `git ls-files -m` since that's the only new file. Unzip it.
4. Add test files and all other essential gadgets to it.
5. Do the test, like the normal CI

**Comment by spacewander** [comment]
> The action would be triggered when the PR is (1) not in master branch (2) with tag release-test

Currently we create tag only after release as an archive. Maybe we should use `release/xxx` branch to trigger the test?

> Upload the package tarball to the PR

We can create the tarball with Makefile. Currently the tarball is created from `VERSION=xxx make release-src`.

**Comment by Yiyiyimu** [comment]
> Currently we create tag only after release as an archive. Maybe we should use release/xxx branch to trigger the test?

Hi @spacewander If I take it right, we create tag after release, but we create branch before the release. So we could use that branch say `v2.3`.

> We can create the tarball with Makefile. Currently the tarball is created from VERSION=xxx make release-src.

Since we could not have bot to add PR, so the tarball created by Github Actions could not be reused by us. And if I take it right, testing the tarball is to avoid problems that happens in manual tarball creation, so we need to test that **exact** tarball we're gonna release.

**Comment by spacewander** [comment]
> My bad, it should be label, not tag

Why not use branch? Since we already need to create a branch.

> Since we could not have bot to add PR, so the tarball created by Github Actions could not be reused by us

I am not quite understand. There are plenty ways to create the tar in the GitHub Actions and transfer it outside. You can upload the tar to any server.

**Comment by Yiyiyimu** [comment]
> Why not use branch? Since we already need to create a branch.

Yes we use both branch and also label. Since there would be PR on the new branch other than the test one.

> Since we could not have bot to add PR, so the tarball created by Github Actions could not be reused by us

I'm glad we could upload it somewhere. So could you tell me where should we upload to
Besides, manually trigger a workflow is only allowed for [master branch](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow#running-a-workflow-on-github). So do you know how could we create that PR to trigger the workflow

**Comment by spacewander** [comment]
> Since there would be PR on the new branch other than the test one

Why should we release a new version based on a PR to non-master branch?

**Comment by Yiyiyimu** [comment]
> Why should we release a new version based on a PR to non-master branch?

Like for now there are 9 commits in branch v2.3 ahead of master

**Comment by spacewander** [comment]
If push to the branch v2.3 can trigger the CI, why should we use label?
Should we label a branch 9 times?

**Comment by Yiyiyimu** [comment]
Ah I get it, you mean we trigger the CI to create the tarball, test it and send it to maintainers each time we push to a release branch. And yeah that could make the release publish much easier, compared to test it in a separate PR.
Also since we would change all the changelog and stuffs in the new branch, so there would be at least one push in the release branch. Cool that's perfect.
